core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2115,6 +2140,39 @@ private[zio] trait ZIOFunctions extends Serializable {
       }
       .refailWithTrace
 
+  /**
+   * Applies the function `f` to each element of the `Chunk[A]` in parallel,
+   * and returns the results in a new `Chunk[B]`.
+   *
+   * For a sequential version of this method, see `foreach`.
+   */
+  final def foreachPar[R, E, A, B](as: Chunk[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, Chunk[B]] = {
+    val len                        = as.length
+    var array: ZIO[R, E, Array[B]] = IO.succeed(null.asInstanceOf[Array[B]])
```

Comment:
Mixing mutation and IO here is bound to create unpredictable behaviour, how about matching the list version with something like:
```scala
as.foldRight(succeed(Chunk.empty))((a, bs) => fn(a).zipWithPar(bs)(_ +: _)).refailWithTrace
```
With the missing `+:`
```scala
def +:[A1 >: A](a: A1): Chunk[A1] = Chunk.single(a) ++ self
```

Reply:
This function matches mapM: https://github.com/zio/zio/blob/6e1960e9223b513cdd86ef5fce69b64c878cbbe3/core/shared/src/main/scala/zio/Chunk.scala#L542

Reply:
😮 It’s baffling that scope is retained as such. My bad, but still I’d suggest either to add an `mapMPar` or use the fold.